dsls/sql/src/main/java/org/apache/beam/dsls/sql/transform/BeamAggregationTransforms.java

```diff
@@ -134,545 +134,201 @@ public Instant apply(BeamSqlRow input) {
             break;
           default:
-            break;
+            throw new UnsupportedOperationException();
           }
           break;
-        case "MIN":
-          switch (ex.getOutputType()) {
+        case "AVG":
+          switch (call.type.getSqlTypeName()) {
           case INTEGER:
-            deltaRecord.addField(idx,
-                Math.min(ex.evaluate(input).getInteger(), accumulator.getInteger(idx)));
-            break;
-          case BIGINT:
-            deltaRecord.addField(idx,
-                Math.min(ex.evaluate(input).getLong(), accumulator.getLong(idx)));
+            aggregators.add(new BeamBuiltinAggregations.Avg<Integer>(outFieldType));
             break;
           case SMALLINT:
-            deltaRecord.addField(idx,
-                (short) Math.min(ex.evaluate(input).getShort(), accumulator.getShort(idx)));
+            aggregators.add(new BeamBuiltinAggregations.Avg<Short>(outFieldType));
             break;
           case TINYINT:
-            deltaRecord.addField(idx,
-                (byte) Math.min(ex.evaluate(input).getByte(), accumulator.getByte(idx)));
+            aggregators.add(new BeamBuiltinAggregations.Avg<Byte>(outFieldType));
+            break;
+          case BIGINT:
+            aggregators.add(new BeamBuiltinAggregations.Avg<Long>(outFieldType));
             break;
           case FLOAT:
-            deltaRecord.addField(idx,
-                Math.min(ex.evaluate(input).getFloat(), accumulator.getFloat(idx)));
+            aggregators.add(new BeamBuiltinAggregations.Avg<Float>(outFieldType));
             break;
           case DOUBLE:
-            deltaRecord.addField(idx,
-                Math.min(ex.evaluate(input).getDouble(), accumulator.getDouble(idx)));
+            aggregators.add(new BeamBuiltinAggregations.Avg<Double>(outFieldType));
             break;
-          case TIMESTAMP:
-            Date preDate = accumulator.getDate(idx);
-            Date nowDate = ex.evaluate(input).getDate();
-            deltaRecord.addField(idx, preDate.getTime() < nowDate.getTime() ? preDate : nowDate);
+          case DECIMAL:
+            aggregators.add(new BeamBuiltinAggregations.Avg<BigDecimal>(outFieldType));
             break;
           default:
-            break;
+            throw new UnsupportedOperationException();
           }
           break;
         default:
-          break;
+          if (call.getAggregation() instanceof SqlUserDefinedAggFunction) {
+            // handle UDAF.
+            SqlUserDefinedAggFunction udaf = (SqlUserDefinedAggFunction) call.getAggregation();
+            AggregateFunctionImpl fn = (AggregateFunctionImpl) udaf.function;
+            try {
+              aggregators.add((BeamSqlUdaf) fn.declaringClass.newInstance());
+            } catch (Exception e) {
+              throw new IllegalStateException(e);
+            }
+          } else {
+            throw new UnsupportedOperationException();
```

Comment:
Provide more details here? This will be a pretty opaque error when it gets hit otherwise.